src/backend/Pg.cpp

```diff
@@ -238,38 +448,87 @@ Pg::bulkInsert(char const* table, std::string const& records)
         throw std::runtime_error(ss.str());
     }
 
-    if (PQputCopyData(conn_.get(), records.c_str(), records.size()) == -1)
+    auto& socket = getSocket();
+
+    std::int32_t putCopy = 0;
+    while (putCopy == 0)
+    {
+        putCopy = PQputCopyData(conn_.get(), records.c_str(), records.size());
+        
+        if (putCopy == -1)
```

Comment:
There are times when we don't want copy to be synchronous. There are multiple tables that all need to be copied to, but if copy is synchronous, we need to wait for one to finish before submitting the next. Also, during the initial ledger load, we flush the buffer via copy every 100,000 or million records or so.

Reply:
Are they all in the same transaction, such as in reporting mode?

Anyway, see the comment above this for Pg.cpp:454. Namely, this function looks broken to me, or at least not working as postgres documents it. Fixing it by making it synchronous would be easier.

Reply:
yeah they are all in the same transaction

Reply:
My understanding is that synchronous `COPY` should work similarly to `query`, but calling `PQputCopyData` or `PQputCopyEnd` instead of `PQsendQueryParams`.

This does _work_, I'm gonna look through the documentation to try to find how to correctly implement it.

Reply:
So after looking at some docs, this looks correct 
" To prevent this possibility and achieve completely nonblocking database operation, the following additional functions can be used.", and then lists `PQFlush`.

We flush after we send each command. 
"After sending any command or data on a nonblocking connection, call PQflush. If it returns 1, wait for the socket to become read- or write-ready. If it becomes write-ready, call PQflush again. If it becomes read-ready, call PQconsumeInput , then call PQflush again. Repeat until PQflush returns 0. (It is necessary to check for read-ready and drain the input with PQconsumeInput , because the server can block trying to send us data, e.g., NOTICE messages, and won't read our data until we read its.) Once PQflush returns 0, wait for the socket to be read-ready and then read the response as described above." 

Using PQflush here should eliminate the issue described above, if I'm RTFM-ing correctly.